Deploy web and task component in independent deployment

TheRealHaoLiu opened the pull request:
##### SUMMARY

Requires https://github.com/ansible/awx/pull/13423 to function
<!--- Describe the change, including rationale and design decisions -->

<!---
If you are fixing an existing issue, please include "fixes #nnn" in your
commit message and your description; but you should still explain what
the change does.
-->

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - New or Enhanced Feature


##### ADDITIONAL INFORMATION
<!---
Include additional information to help people understand the change here.
For bugs that don't have a linked bug report, a step-by-step reproduction
of the problem is helpful.
-->

<!--- Paste verbatim command output below, e.g. before and after your change -->
```

```

shanemcd committed:
Split web and task containers into separate deployments

relrod committed:
drop awx-web container from awx-task deployment

Signed-off-by: Rick Elrod <rick@elrod.me>

relrod committed:
add $MY_POD_IP back to web deployment

Signed-off-by: Rick Elrod <rick@elrod.me>

relrod committed:
scale web/task replicas separately

This is a breaking change since 'replicas' is no longer used.

Signed-off-by: Rick Elrod <rick@elrod.me>

thedoubl3j committed:
[web/task split] split web and task deployment + a few supporting bits (#1189)

* first pass, still WIP, need tolerations etc

* add tolerations that don't work bc idk

* bug hunting

* local push, still a WIP

* affinity still needs testfor to_nice_yaml, tolerations logic is working

* fixed task deployment and affinity for both

thedoubl3j committed:
add nodeSelector work (#1219)

thedoubl3j committed:
fix affinity duplicate (#1225)

TheRealHaoLiu committed:
Add web and task replicas to the CRD  (#1227)

add web_replicas and task_replicas

Co-Authored-By: Jake Jackson  <24478650+thedoubl3j@users.noreply.github.com>

thedoubl3j committed:
add topology constraint for each deployment (#1234)

jessicamack committed:
added descriptors for new fields (#1238)

Signed-off-by: jessicamack <jmack@redhat.com>

thedoubl3j committed:
update documentation for the new fields added

TheRealHaoLiu committed:
update auto_upgrade logic (#1241)

update logic for determining if install.yml task should be run
to respect the auto_upgrade field in awx resource

conditions and expected behavior
```
  auto_upgrade   awx   awx-web   awx-task   run install.yml
 -------------- ----- --------- ---------- -----------------
  T              -     -         -          T
  F              T     -         -          F
  F              -     T         T          F
  F              -     T         F          T
  F              -     F         T          T
  F              -     F         F          T
```

thedoubl3j committed:
remove old deployment during upgrade

TheRealHaoLiu committed:
enable web container to be restarted when configmap/secret change

TheRealHaoLiu committed:
rename tower_pod to awx_task_pod

TheRealHaoLiu committed:
resolve merge error

thedoubl3j committed:
remove any reference to project persistance in web deployment

djyasin reviewed (commented):
Did not find any references to `tower_pod` in revised code

thedoubl3j reviewed (commented):
for now I think this looks good, I am not seeing anything immediately on my first pass that would be a cause for concern. I will continue going through current code base to make sure we are renaming everything to the new names etc.

djyasin committed:
Updated task deployment to fix receptor certs issue.
Added the following volume mounts to the web container:
- receptor-work-signing
- receptor-ca
- work-public-key.pem
Also added these corresponding volumes to the web deployments:
- receptor-ca
- receptor-work-signing

TheRealHaoLiu committed:
add AWX_COMPONENT env var

for awx to be able to identify which component it is running in code

thedoubl3j committed:
[web/task split] fix scale down bug  (#1295)

- rename scale_down vars to the new deployments since the old one no longer exists
- rename postgres.yml scale down vars as it references the old ones as well

thedoubl3j commented:
the failure is caused by this 
```
"status": {
              "conditions": [
                  {
                      "lastTransitionTime": "2023-03-28T15:03:37Z",
                      "reason": "",
                      "status": "False",
                      "type": "Successful"
                  },
                  {
                      "lastTransitionTime": "2023-03-28T15:13:12Z",
                      "reason": "Failed",
                      "status": "False",
                      "type": "Failure"
                  },
                  {
                      "lastTransitionTime": "2023-03-28T15:13:12Z",
                      "reason": "Running",
                      "status": "True",
```
which we were able to reproduce locally. 

This is caused by us not being able to override the default image and image_version for the molecule deployment. By default, those are set to devel which will fail molecule testing because the new launch script for the rsyslog sidecar is not found in the devel image, but rather the feature image for this new split feature. 

#1308 has been opened to fix this. we were able to fix this locally by just swapping the hardcoded image and image_version.

TheRealHaoLiu commented on Makefile:
```diff
@@ -67,6 +67,9 @@ CHART_REPO ?= awx-operator
 CHART_BRANCH ?= gh-pages
 CHART_INDEX ?= index.yaml
 
+# Development
```
this shouldnt be in this PR lets make the PR against devel

djyasin replied:
Just to confirm, do we want to cut out these commits and put them in a separate PR?

TheRealHaoLiu commented on Makefile:
```diff
@@ -87,6 +90,11 @@ all: docker-build
 help: ## Display this help.
 	@awk 'BEGIN {FS = ":.*##"; printf "\nUsage:\n  make \033[36m<target>\033[0m\n"} /^[a-zA-Z_0-9-]+:.*?##/ { printf "  \033[36m%-15s\033[0m %s\n", $$1, $$2 } /^##@/ { printf "\n\033[1m%s\033[0m\n", substr($$0, 5) } ' $(MAKEFILE_LIST)
 
+##@ Development
```
this should be in a separate PR

TheRealHaoLiu commented on roles/installer/defaults/main.yml:
```diff
@@ -72,6 +72,22 @@ hostname: ''
+# Add a nodeSelector for the AWX pods. It must match a node's labels for the pod
+# to be scheduled on that node. Specify as literal block. E.g.:
+# node_selector: |
+#   disktype: ssd
```
```suggestion
```

TheRealHaoLiu commented on roles/installer/defaults/main.yml:
```diff
@@ -72,6 +72,22 @@ hostname: ''
+# Add a nodeSelector for the AWX pods. It must match a node's labels for the pod
+# to be scheduled on that node. Specify as literal block. E.g.:
+# node_selector: |
+#   disktype: ssd
```
```suggestion
```

TheRealHaoLiu commented on roles/installer/defaults/main.yml:
```diff
@@ -83,7 +99,27 @@ node_selector: ''
+#     whenUnsatisfiable: "ScheduleAnyway"
+#     labelSelector:
+#       matchLabels:
+#         app.kubernetes.io/name: "<resourcename>""
```
```suggestion
#         app.kubernetes.io/name: "<resourcename>"
```

djyasin replied:
Resolved in https://github.com/ansible/awx-operator/pull/1311

TheRealHaoLiu commented on roles/installer/templates/deployments/task.yaml.j2:
```diff
@@ -340,6 +238,8 @@ spec:
                 - /var/lib/pre-stop/scripts/termination-master
 {% endif %}
           env:
+            - name: AWX_COMPONENT
```
this is added so that awx get_task_qeuename() function can identify when in web container and will try to select a task instance's queue

TheRealHaoLiu commented on roles/installer/templates/networking/service.yaml.j2:
```diff
@@ -13,7 +13,7 @@ metadata:
 spec:
   ports:
-  
+
```
```suggestion
```

djyasin replied:
Resolve in https://github.com/ansible/awx-operator/pull/1311

djyasin committed:
Removed errant quotes in main.yml.

djyasin committed:
Removed extra space.

TheRealHaoLiu committed:
Update arg for launching web container

related to https://github.com/ansible/awx/pull/13777

make sure that the launch script name reflect the component that its launching

TheRealHaoLiu committed:
Rename SUPERVISOR_WEB_CONFIG_PATH

SUPERVISOR_WEB_CONFIG_PATH is used in the old deployment for task container to reach into the web container and restart services

this is no longer possible/needed after splitting the deployment

renaming SUPERVISOR_WEB_CONFIG_PATH to SUPERVISOR_CONFIG_PATH
and setting it to the supervisor file for the container

this can still be useful to help run `supervisorctl -c $SUPERVISOR_CONFIG_PATH`

thedoubl3j committed:
remove disktype key from example
